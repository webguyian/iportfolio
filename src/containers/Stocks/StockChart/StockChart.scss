.stock-chart {
  &-controls {
    @include unstyled-list;
    display: flex;

    .ui-btn {
      padding: 8px;
      border-radius: 4px;
      color: $white;
    }
  }

  &-control {
    margin: 0 $gutter;

    &--active {
      > .ui-btn {
        background-color: $gray-medium;
      }
    }
  }
}
