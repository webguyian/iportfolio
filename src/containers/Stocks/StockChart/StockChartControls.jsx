import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Button from 'components/Button/Button';

const StockChartControls = props => {
  const { active, onClick, options } = props;
  const baseClass = 'stock-chart-controls';
  const optionClass = 'stock-chart-control';
  const activeClass = `${optionClass}--active`;

  return (
    <ul className={baseClass}>
      {options.map(control => (
        <li
          key={control}
          className={classNames(optionClass, control === active && activeClass)}
        >
          <Button onClick={onClick.bind(null, control)}>{control}</Button>
        </li>
      ))}
    </ul>
  );
};

StockChartControls.propTypes = {
  active: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  options: PropTypes.array.isRequired
};

export default StockChartControls;
