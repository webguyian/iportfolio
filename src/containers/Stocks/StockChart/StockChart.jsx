import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { ChartCanvas, Chart } from 'react-stockcharts';
import { XAxis, YAxis } from 'react-stockcharts/lib/axes';
import { AreaSeries } from 'react-stockcharts/lib/series';
import { scaleTime } from 'd3-scale';

import { formatDateTick } from 'modules/stocks/helpers';
import { useStockChart } from 'modules/stocks/hooks';

import StockChartControls from './StockChartControls';

const StockChart = props => {
  const { height, hideControls, ratio, stock, width, ...otherProps } = props;
  const symbol = stock.symbol;
  const [range, setRange] = useState('1D');
  const chartData = useStockChart(symbol, range, stock.chartData);
  const baseClass = 'stock-chart';
  const showControls = !hideControls;
  const gradientId = showControls
    ? `gradient-${symbol}`
    : `gradient-${symbol}-no-controls`;

  if (!chartData) {
    return null;
  }

  const diffColor = chartData.start < chartData.end ? '#6eb26d' : '#e94741';
  const dateAccessor = d => d.date;

  return (
    <div className={baseClass}>
      {showControls && (
        <StockChartControls
          active={range}
          options={['1D', '1W', '1M', '3M', '6M']}
          onClick={setRange}
        />
      )}
      <ChartCanvas
        data={chartData.data}
        width={width}
        height={height}
        margin={{ left: 15, right: 50, top: 10, bottom: 50 }}
        ratio={ratio}
        seriesName={symbol}
        type="svg"
        xAccessor={dateAccessor}
        displayXAccessor={dateAccessor}
        xScale={scaleTime()}
        xExtents={[chartData.startDate, chartData.endDate]}
        {...otherProps}
      >
        <Chart id={chartData.end} yExtents={d => d.close}>
          <defs>
            <linearGradient id={gradientId} x1="0" y1="100%" x2="0" y2="0%">
              <stop offset="0%" stopColor={diffColor} stopOpacity={0.2} />
              <stop offset="70%" stopColor={diffColor} stopOpacity={0.4} />
              <stop offset="100%" stopColor={diffColor} stopOpacity={0.8} />
            </linearGradient>
          </defs>
          {showControls && (
            <XAxis
              axisAt="bottom"
              orient="bottom"
              ticks={4}
              tickStroke="#FFFFFF"
              tickFormat={formatDateTick.bind(null, chartData.range)}
            />
          )}
          {showControls && (
            <YAxis
              axisAt="right"
              orient="right"
              ticks={3}
              tickStroke="#FFFFFF"
              tickFormat={tick => `${Math.round(tick)}`}
            />
          )}
          <AreaSeries
            stroke={diffColor}
            fill={`url(#${gradientId})`}
            yAccessor={d => d.close}
          />
        </Chart>
      </ChartCanvas>
    </div>
  );
};

StockChart.propTypes = {
  data: PropTypes.array,
  height: PropTypes.number,
  hideControls: PropTypes.bool,
  ratio: PropTypes.number,
  width: PropTypes.number,
  stock: PropTypes.object
};

StockChart.defaultProps = {
  height: 200,
  hideControls: false,
  ratio: 1,
  width: 345
};

export default StockChart;
