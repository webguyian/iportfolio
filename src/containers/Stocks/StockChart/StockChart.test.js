import React from 'react';
import { act, create } from 'react-test-renderer';

import StockChart from './StockChart';

describe('<StockChart />', () => {
  // Override prop warning from react-stockcharts
  console.warn = jest.fn();
  const stock = {
    c: 183.63,
    h: 183.74,
    l: 183.5,
    o: 183.7,
    pc: 183.7,
    symbol: 'MSFT',
    displaySymbol: 'MSFT',
    description: 'MICROSOFT CORP',
    price: 188.7,
    previousPrice: 183.89,
    chartData: {
      data: [
        {
          date: '2020-02-10T09:00:00.000Z',
          timestamp: 1581325200,
          close: 184
        },
        {
          date: '2020-02-10T09:03:00.000Z',
          timestamp: 1581325380,
          close: 183.96
        },
        {
          date: '2020-02-10T09:04:00.000Z',
          timestamp: 1581325440,
          close: 184.1
        },
        {
          date: '2020-02-10T09:12:00.000Z',
          timestamp: 1581325920,
          close: 184.23
        },
        {
          date: '2020-02-10T09:16:00.000Z',
          timestamp: 1581326160,
          close: 184.44
        },
        {
          date: '2020-02-10T09:18:00.000Z',
          timestamp: 1581326280,
          close: 184.26
        },
        {
          date: '2020-02-10T09:20:00.000Z',
          timestamp: 1581326400,
          close: 184.33
        },
        {
          date: '2020-02-10T09:37:00.000Z',
          timestamp: 1581327420,
          close: 184.35
        },
        {
          date: '2020-02-10T09:43:00.000Z',
          timestamp: 1581327780,
          close: 184.2
        },
        {
          date: '2020-02-10T09:45:00.000Z',
          timestamp: 1581327900,
          close: 184.1
        }
      ],
      start: 184,
      startDate: '2020-02-10T09:00:00.000Z',
      end: 189.68,
      endDate: '2020-02-11T00:59:00.000Z',
      range: '1D'
    }
  };

  it('renders correctly', async () => {
    let component;

    await act(async () => {
      component = create(<StockChart stock={stock} />);
    });

    expect(component).toMatchSnapshot();
  });

  it('renders correctly without controls', async () => {
    let component;

    await act(async () => {
      component = create(<StockChart stock={stock} hideControls />);
    });

    expect(component).toMatchSnapshot();
  });

  it('renders correctly without chartData', async () => {
    const noData = {
      ...stock,
      chartData: null
    };

    let component;

    await act(async () => {
      component = create(<StockChart stock={noData} hideControls />);
    });

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with downgrade', async () => {
    const downStock = {
      ...stock,
      chartData: {
        ...stock.chartData,
        end: 180
      }
    };

    let component;

    await act(async () => {
      component = create(<StockChart stock={downStock} hideControls />);
    });

    expect(component).toMatchSnapshot();
  });
});
