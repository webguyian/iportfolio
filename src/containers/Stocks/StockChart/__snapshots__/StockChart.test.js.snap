// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StockChart /> renders correctly 1`] = `
<div
  className="stock-chart"
>
  <ul
    className="stock-chart-controls"
  >
    <li
      className="stock-chart-control stock-chart-control--active"
    >
      <button
        className="ui-btn"
        onClick={[Function]}
        type="button"
      >
        1D
      </button>
    </li>
    <li
      className="stock-chart-control"
    >
      <button
        className="ui-btn"
        onClick={[Function]}
        type="button"
      >
        1W
      </button>
    </li>
    <li
      className="stock-chart-control"
    >
      <button
        className="ui-btn"
        onClick={[Function]}
        type="button"
      >
        1M
      </button>
    </li>
    <li
      className="stock-chart-control"
    >
      <button
        className="ui-btn"
        onClick={[Function]}
        type="button"
      >
        3M
      </button>
    </li>
    <li
      className="stock-chart-control"
    >
      <button
        className="ui-btn"
        onClick={[Function]}
        type="button"
      >
        6M
      </button>
    </li>
  </ul>
  <div
    className="react-stockchart"
    onClick={[Function]}
    style={
      Object {
        "height": 200,
        "position": "relative",
        "width": 345,
      }
    }
  >
    <svg
      className="react-stockchart"
      height={200}
      style={
        Object {
          "position": "absolute",
          "zIndex": 6,
        }
      }
      width={345}
    >
      <style
        type="text/css"
      >
        
	.react-stockcharts-grabbing-cursor {
		pointer-events: all;
		cursor: -moz-grabbing;
		cursor: -webkit-grabbing;
		cursor: grabbing;
	}
	.react-stockcharts-crosshair-cursor {
		pointer-events: all;
		cursor: crosshair;
	}
	.react-stockcharts-tooltip-hover {
		pointer-events: all;
		cursor: pointer;
	}
	.react-stockcharts-avoid-interaction {
		pointer-events: none;
	}
	.react-stockcharts-enable-interaction {
		pointer-events: all;
	}
	.react-stockcharts-tooltip {
		pointer-events: all;
		cursor: pointer;
	}
	.react-stockcharts-default-cursor {
		cursor: default;
	}
	.react-stockcharts-move-cursor {
		cursor: move;
	}
	.react-stockcharts-pointer-cursor {
		cursor: pointer;
	}
	.react-stockcharts-ns-resize-cursor {
		cursor: ns-resize;
	}
	.react-stockcharts-ew-resize-cursor {
		cursor: ew-resize;
	}
      </style>
      <defs>
        <clipPath
          id="chart-area-clip"
        >
          <rect
            height={140}
            width={280}
            x="0"
            y="0"
          />
        </clipPath>
        <clipPath
          id="chart-area-clip-189.68"
        >
          <rect
            height={140}
            width={280}
            x="0"
            y="0"
          />
        </clipPath>
      </defs>
      <g
        transform="translate(15.5, 10.5)"
      >
        <rect
          className=""
          height={140}
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          onWheel={[Function]}
          style={
            Object {
              "opacity": 0,
            }
          }
          width={280}
        />
        <g
          className="react-stockcharts-avoid-interaction"
        >
          <g
            transform="translate(0, 0)"
          >
            <defs>
              <linearGradient
                id="gradient-MSFT"
                x1="0"
                x2="0"
                y1="100%"
                y2="0%"
              >
                <stop
                  offset="0%"
                  stopColor="#6eb26d"
                  stopOpacity={0.2}
                />
                <stop
                  offset="70%"
                  stopColor="#6eb26d"
                  stopOpacity={0.4}
                />
                <stop
                  offset="100%"
                  stopColor="#6eb26d"
                  stopOpacity={0.8}
                />
              </linearGradient>
            </defs>
            <g
              transform="translate(0, 140)"
            >
              <rect
                className="react-stockcharts-enable-interaction react-stockcharts-default-cursor react-stockcharts-x-axis"
                height={25}
                onContextMenu={[Function]}
                onMouseDown={[Function]}
                onTouchStart={[Function]}
                opacity={0}
                width={280}
                x={0}
                y={0}
              />
              <g
                style={null}
              >
                <g
                  className="react-stockcharts-x-axis"
                >
                  <g />
                  <path
                    className="react-stockcharts-axis-domain"
                    d="M0,0V0H280V0"
                    fill="none"
                    opacity={1}
                    stroke="#000000"
                    strokeWidth={1}
                  />
                </g>
              </g>
            </g>
            <g
              transform="translate(280, 0)"
            >
              <rect
                className="react-stockcharts-enable-interaction react-stockcharts-default-cursor react-stockcharts-y-axis"
                height={140}
                onContextMenu={[Function]}
                onMouseDown={[Function]}
                onTouchStart={[Function]}
                opacity={0}
                width={40}
                x={0}
                y={0}
              />
              <g
                style={null}
              >
                <g
                  className="react-stockcharts-y-axis"
                >
                  <g>
                    <g
                      className="tick"
                    >
                      <line
                        opacity={1}
                        shapeRendering="crispEdges"
                        stroke="#FFFFFF"
                        strokeDasharray="none"
                        x1={0}
                        x2={5}
                        y1={128}
                        y2={128}
                      />
                      <text
                        dy=".32em"
                        fill="#FFFFFF"
                        fontFamily="Helvetica Neue, Helvetica, Arial, sans-serif"
                        fontSize={12}
                        fontWeight={400}
                        textAnchor="start"
                        x={11}
                        y={128}
                      >
                        184
                      </text>
                    </g>
                    <g
                      className="tick"
                    >
                      <line
                        opacity={1}
                        shapeRendering="crispEdges"
                        stroke="#FFFFFF"
                        strokeDasharray="none"
                        x1={0}
                        x2={5}
                        y1={70}
                        y2={70}
                      />
                      <text
                        dy=".32em"
                        fill="#FFFFFF"
                        fontFamily="Helvetica Neue, Helvetica, Arial, sans-serif"
                        fontSize={12}
                        fontWeight={400}
                        textAnchor="start"
                        x={11}
                        y={70}
                      >
                        184
                      </text>
                    </g>
                    <g
                      className="tick"
                    >
                      <line
                        opacity={1}
                        shapeRendering="crispEdges"
                        stroke="#FFFFFF"
                        strokeDasharray="none"
                        x1={0}
                        x2={5}
                        y1={12}
                        y2={12}
                      />
                      <text
                        dy=".32em"
                        fill="#FFFFFF"
                        fontFamily="Helvetica Neue, Helvetica, Arial, sans-serif"
                        fontSize={12}
                        fontWeight={400}
                        textAnchor="start"
                        x={11}
                        y={12}
                      >
                        184
                      </text>
                    </g>
                  </g>
                  <path
                    className="react-stockcharts-axis-domain"
                    d="M0,0H0V140H0"
                    fill="none"
                    opacity={1}
                    stroke="#FFFFFF"
                    strokeWidth={1}
                  />
                </g>
              </g>
            </g>
            <g
              className="react-stockcharts-area"
            >
              <g
                style={
                  Object {
                    "clipPath": "url(#chart-area-clip-189.68)",
                  }
                }
              >
                <path
                  className="line "
                  d="MNaN,128LNaN,140LNaN,99LNaN,61LNaN,0LNaN,53LNaN,32LNaN,26LNaN,70LNaN,99LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140Z"
                  fill="url(#gradient-MSFT)"
                  stroke="none"
                />
              </g>
              <g
                style={
                  Object {
                    "clipPath": "url(#chart-area-clip-189.68)",
                  }
                }
              >
                <path
                  className="line  "
                  d="MNaN,128LNaN,140LNaN,99LNaN,61LNaN,0LNaN,53LNaN,32LNaN,26LNaN,70LNaN,99"
                  fill="none"
                  stroke="#6eb26d"
                  strokeDasharray="none"
                  strokeOpacity={1}
                  strokeWidth={1}
                />
              </g>
            </g>
          </g>
        </g>
      </g>
    </svg>
  </div>
</div>
`;

exports[`<StockChart /> renders correctly with downgrade 1`] = `
<div
  className="stock-chart"
>
  <div
    className="react-stockchart"
    onClick={[Function]}
    style={
      Object {
        "height": 200,
        "position": "relative",
        "width": 345,
      }
    }
  >
    <svg
      className="react-stockchart"
      height={200}
      style={
        Object {
          "position": "absolute",
          "zIndex": 6,
        }
      }
      width={345}
    >
      <style
        type="text/css"
      >
        
	.react-stockcharts-grabbing-cursor {
		pointer-events: all;
		cursor: -moz-grabbing;
		cursor: -webkit-grabbing;
		cursor: grabbing;
	}
	.react-stockcharts-crosshair-cursor {
		pointer-events: all;
		cursor: crosshair;
	}
	.react-stockcharts-tooltip-hover {
		pointer-events: all;
		cursor: pointer;
	}
	.react-stockcharts-avoid-interaction {
		pointer-events: none;
	}
	.react-stockcharts-enable-interaction {
		pointer-events: all;
	}
	.react-stockcharts-tooltip {
		pointer-events: all;
		cursor: pointer;
	}
	.react-stockcharts-default-cursor {
		cursor: default;
	}
	.react-stockcharts-move-cursor {
		cursor: move;
	}
	.react-stockcharts-pointer-cursor {
		cursor: pointer;
	}
	.react-stockcharts-ns-resize-cursor {
		cursor: ns-resize;
	}
	.react-stockcharts-ew-resize-cursor {
		cursor: ew-resize;
	}
      </style>
      <defs>
        <clipPath
          id="chart-area-clip"
        >
          <rect
            height={140}
            width={280}
            x="0"
            y="0"
          />
        </clipPath>
        <clipPath
          id="chart-area-clip-180"
        >
          <rect
            height={140}
            width={280}
            x="0"
            y="0"
          />
        </clipPath>
      </defs>
      <g
        transform="translate(15.5, 10.5)"
      >
        <rect
          className=""
          height={140}
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          onWheel={[Function]}
          style={
            Object {
              "opacity": 0,
            }
          }
          width={280}
        />
        <g
          className="react-stockcharts-avoid-interaction"
        >
          <g
            transform="translate(0, 0)"
          >
            <defs>
              <linearGradient
                id="gradient-MSFT-no-controls"
                x1="0"
                x2="0"
                y1="100%"
                y2="0%"
              >
                <stop
                  offset="0%"
                  stopColor="#e94741"
                  stopOpacity={0.2}
                />
                <stop
                  offset="70%"
                  stopColor="#e94741"
                  stopOpacity={0.4}
                />
                <stop
                  offset="100%"
                  stopColor="#e94741"
                  stopOpacity={0.8}
                />
              </linearGradient>
            </defs>
            <g
              className="react-stockcharts-area"
            >
              <g
                style={
                  Object {
                    "clipPath": "url(#chart-area-clip-180)",
                  }
                }
              >
                <path
                  className="line "
                  d="MNaN,128LNaN,140LNaN,99LNaN,61LNaN,0LNaN,53LNaN,32LNaN,26LNaN,70LNaN,99LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140Z"
                  fill="url(#gradient-MSFT-no-controls)"
                  stroke="none"
                />
              </g>
              <g
                style={
                  Object {
                    "clipPath": "url(#chart-area-clip-180)",
                  }
                }
              >
                <path
                  className="line  "
                  d="MNaN,128LNaN,140LNaN,99LNaN,61LNaN,0LNaN,53LNaN,32LNaN,26LNaN,70LNaN,99"
                  fill="none"
                  stroke="#e94741"
                  strokeDasharray="none"
                  strokeOpacity={1}
                  strokeWidth={1}
                />
              </g>
            </g>
          </g>
        </g>
      </g>
    </svg>
  </div>
</div>
`;

exports[`<StockChart /> renders correctly without chartData 1`] = `null`;

exports[`<StockChart /> renders correctly without controls 1`] = `
<div
  className="stock-chart"
>
  <div
    className="react-stockchart"
    onClick={[Function]}
    style={
      Object {
        "height": 200,
        "position": "relative",
        "width": 345,
      }
    }
  >
    <svg
      className="react-stockchart"
      height={200}
      style={
        Object {
          "position": "absolute",
          "zIndex": 6,
        }
      }
      width={345}
    >
      <style
        type="text/css"
      >
        
	.react-stockcharts-grabbing-cursor {
		pointer-events: all;
		cursor: -moz-grabbing;
		cursor: -webkit-grabbing;
		cursor: grabbing;
	}
	.react-stockcharts-crosshair-cursor {
		pointer-events: all;
		cursor: crosshair;
	}
	.react-stockcharts-tooltip-hover {
		pointer-events: all;
		cursor: pointer;
	}
	.react-stockcharts-avoid-interaction {
		pointer-events: none;
	}
	.react-stockcharts-enable-interaction {
		pointer-events: all;
	}
	.react-stockcharts-tooltip {
		pointer-events: all;
		cursor: pointer;
	}
	.react-stockcharts-default-cursor {
		cursor: default;
	}
	.react-stockcharts-move-cursor {
		cursor: move;
	}
	.react-stockcharts-pointer-cursor {
		cursor: pointer;
	}
	.react-stockcharts-ns-resize-cursor {
		cursor: ns-resize;
	}
	.react-stockcharts-ew-resize-cursor {
		cursor: ew-resize;
	}
      </style>
      <defs>
        <clipPath
          id="chart-area-clip"
        >
          <rect
            height={140}
            width={280}
            x="0"
            y="0"
          />
        </clipPath>
        <clipPath
          id="chart-area-clip-189.68"
        >
          <rect
            height={140}
            width={280}
            x="0"
            y="0"
          />
        </clipPath>
      </defs>
      <g
        transform="translate(15.5, 10.5)"
      >
        <rect
          className=""
          height={140}
          onClick={[Function]}
          onContextMenu={[Function]}
          onMouseDown={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          onWheel={[Function]}
          style={
            Object {
              "opacity": 0,
            }
          }
          width={280}
        />
        <g
          className="react-stockcharts-avoid-interaction"
        >
          <g
            transform="translate(0, 0)"
          >
            <defs>
              <linearGradient
                id="gradient-MSFT-no-controls"
                x1="0"
                x2="0"
                y1="100%"
                y2="0%"
              >
                <stop
                  offset="0%"
                  stopColor="#6eb26d"
                  stopOpacity={0.2}
                />
                <stop
                  offset="70%"
                  stopColor="#6eb26d"
                  stopOpacity={0.4}
                />
                <stop
                  offset="100%"
                  stopColor="#6eb26d"
                  stopOpacity={0.8}
                />
              </linearGradient>
            </defs>
            <g
              className="react-stockcharts-area"
            >
              <g
                style={
                  Object {
                    "clipPath": "url(#chart-area-clip-189.68)",
                  }
                }
              >
                <path
                  className="line "
                  d="MNaN,128LNaN,140LNaN,99LNaN,61LNaN,0LNaN,53LNaN,32LNaN,26LNaN,70LNaN,99LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140LNaN,140Z"
                  fill="url(#gradient-MSFT-no-controls)"
                  stroke="none"
                />
              </g>
              <g
                style={
                  Object {
                    "clipPath": "url(#chart-area-clip-189.68)",
                  }
                }
              >
                <path
                  className="line  "
                  d="MNaN,128LNaN,140LNaN,99LNaN,61LNaN,0LNaN,53LNaN,32LNaN,26LNaN,70LNaN,99"
                  fill="none"
                  stroke="#6eb26d"
                  strokeDasharray="none"
                  strokeOpacity={1}
                  strokeWidth={1}
                />
              </g>
            </g>
          </g>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
